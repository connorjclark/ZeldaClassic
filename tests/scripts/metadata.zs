/** COMMENT IGNORED */
// COMMENT IGNORED
int postfix = 1; // THIS COMMENT USED
/** x marks the spot */
int x = 1;

// A car, duh
class Car {
    /** How fast it goes */
    int speed;

	/** Go fast */
    void vroom() {
        speed = 1;
    }

    void vroom2() {
        vroom();
    }
}

// a ffc script called Metadata
ffc script Metadata {
    // run me
    void run(int radius, int speed, int angle, int radius2, int angle2) {
        printf("%d %d\n", postfix, 1 + x + 2);
        utils::fn(postfix);
        // i am a car
        auto c = new Car();
        printf("%d %d %d\n", c->speed, radius, utils::hmm);
    }
}

// a namespace for cool things
namespace utils {
    /** A lovely function. */
    void fn(int a) {

    }

    int hmm; // hmmmmmm
}
