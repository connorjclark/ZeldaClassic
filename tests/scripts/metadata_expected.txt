Compiling 'metadata.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Checking code paths
Pass 6: Generating object code
Pass 7: Assembling
Success!
Compile finished with exit code '0' (success)
=== METADATA
{
  "currentFileSymbols": [
    {
      "name": "Fancy",
      "kind": 13,
      "range": {
        "start": {
          "line": 46,
          "character": 4
        },
        "end": {
          "line": 46,
          "character": 9
        }
      },
      "selectionRange": {
        "start": {
          "line": 46,
          "character": 4
        },
        "end": {
          "line": 46,
          "character": 9
        }
      },
      "children": []
    },
    {
      "name": "Dull",
      "kind": 13,
      "range": {
        "start": {
          "line": 48,
          "character": 4
        },
        "end": {
          "line": 48,
          "character": 8
        }
      },
      "selectionRange": {
        "start": {
          "line": 48,
          "character": 4
        },
        "end": {
          "line": 48,
          "character": 8
        }
      },
      "children": []
    },
    {
      "name": "Mid",
      "kind": 13,
      "range": {
        "start": {
          "line": 49,
          "character": 4
        },
        "end": {
          "line": 49,
          "character": 7
        }
      },
      "selectionRange": {
        "start": {
          "line": 49,
          "character": 4
        },
        "end": {
          "line": 49,
          "character": 7
        }
      },
      "children": []
    },
    {
      "name": "postfix",
      "kind": 13,
      "range": {
        "start": {
          "line": 2,
          "character": 4
        },
        "end": {
          "line": 2,
          "character": 15
        }
      },
      "selectionRange": {
        "start": {
          "line": 2,
          "character": 4
        },
        "end": {
          "line": 2,
          "character": 15
        }
      },
      "children": []
    },
    {
      "name": "postfix2",
      "kind": 13,
      "range": {
        "start": {
          "line": 3,
          "character": 4
        },
        "end": {
          "line": 3,
          "character": 22
        }
      },
      "selectionRange": {
        "start": {
          "line": 3,
          "character": 4
        },
        "end": {
          "line": 3,
          "character": 22
        }
      },
      "children": []
    },
    {
      "name": "x",
      "kind": 13,
      "range": {
        "start": {
          "line": 16,
          "character": 4
        },
        "end": {
          "line": 16,
          "character": 9
        }
      },
      "selectionRange": {
        "start": {
          "line": 16,
          "character": 4
        },
        "end": {
          "line": 16,
          "character": 9
        }
      },
      "children": []
    },
    {
      "name": "y",
      "kind": 13,
      "range": {
        "start": {
          "line": 18,
          "character": 4
        },
        "end": {
          "line": 18,
          "character": 5
        }
      },
      "selectionRange": {
        "start": {
          "line": 18,
          "character": 4
        },
        "end": {
          "line": 18,
          "character": 5
        }
      },
      "children": []
    },
    {
      "name": "z",
      "kind": 13,
      "range": {
        "start": {
          "line": 18,
          "character": 7
        },
        "end": {
          "line": 18,
          "character": 12
        }
      },
      "selectionRange": {
        "start": {
          "line": 18,
          "character": 7
        },
        "end": {
          "line": 18,
          "character": 12
        }
      },
      "children": []
    },
    {
      "name": "utils",
      "kind": 3,
      "range": {
        "start": {
          "line": 67,
          "character": 0
        },
        "end": {
          "line": 74,
          "character": 1
        }
      },
      "selectionRange": {
        "start": {
          "line": 67,
          "character": 0
        },
        "end": {
          "line": 74,
          "character": 1
        }
      },
      "children": [
        {
          "name": "hmm",
          "kind": 13,
          "range": {
            "start": {
              "line": 73,
              "character": 8
            },
            "end": {
              "line": 73,
              "character": 11
            }
          },
          "selectionRange": {
            "start": {
              "line": 73,
              "character": 8
            },
            "end": {
              "line": 73,
              "character": 11
            }
          },
          "children": []
        },
        {
          "name": "fn",
          "kind": 12,
          "range": {
            "start": {
              "line": 69,
              "character": 9
            },
            "end": {
              "line": 69,
              "character": 18
            }
          },
          "selectionRange": {
            "start": {
              "line": 69,
              "character": 9
            },
            "end": {
              "line": 69,
              "character": 18
            }
          },
          "children": []
        }
      ]
    },
    {
      "name": "Metadata",
      "kind": 2,
      "range": {
        "start": {
          "line": 53,
          "character": 0
        },
        "end": {
          "line": 64,
          "character": 1
        }
      },
      "selectionRange": {
        "start": {
          "line": 53,
          "character": 11
        },
        "end": {
          "line": 53,
          "character": 19
        }
      },
      "children": [
        {
          "name": "run",
          "kind": 12,
          "range": {
            "start": {
              "line": 55,
              "character": 9
            },
            "end": {
              "line": 55,
              "character": 71
            }
          },
          "selectionRange": {
            "start": {
              "line": 55,
              "character": 9
            },
            "end": {
              "line": 55,
              "character": 71
            }
          },
          "children": [
            {
              "name": "c",
              "kind": 13,
              "range": {
                "start": {
                  "line": 59,
                  "character": 13
                },
                "end": {
                  "line": 59,
                  "character": 27
                }
              },
              "selectionRange": {
                "start": {
                  "line": 59,
                  "character": 13
                },
                "end": {
                  "line": 59,
                  "character": 27
                }
              },
              "children": []
            },
            {
              "name": "d",
              "kind": 13,
              "range": {
                "start": {
                  "line": 61,
                  "character": 13
                },
                "end": {
                  "line": 61,
                  "character": 36
                }
              },
              "selectionRange": {
                "start": {
                  "line": 61,
                  "character": 13
                },
                "end": {
                  "line": 61,
                  "character": 36
                }
              },
              "children": []
            },
            {
              "name": "c2",
              "kind": 13,
              "range": {
                "start": {
                  "line": 62,
                  "character": 12
                },
                "end": {
                  "line": 62,
                  "character": 27
                }
              },
              "selectionRange": {
                "start": {
                  "line": 62,
                  "character": 12
                },
                "end": {
                  "line": 62,
                  "character": 27
                }
              },
              "children": []
            }
          ]
        }
      ]
    },
    {
      "name": "Car",
      "kind": 5,
      "range": {
        "start": {
          "line": 21,
          "character": 0
        },
        "end": {
          "line": 38,
          "character": 1
        }
      },
      "selectionRange": {
        "start": {
          "line": 21,
          "character": 6
        },
        "end": {
          "line": 21,
          "character": 9
        }
      },
      "children": [
        {
          "name": "speed",
          "kind": 13,
          "range": {
            "start": {
              "line": 23,
              "character": 8
            },
            "end": {
              "line": 23,
              "character": 13
            }
          },
          "selectionRange": {
            "start": {
              "line": 23,
              "character": 8
            },
            "end": {
              "line": 23,
              "character": 13
            }
          },
          "children": []
        },
        {
          "name": "vroom",
          "kind": 12,
          "range": {
            "start": {
              "line": 31,
              "character": 9
            },
            "end": {
              "line": 31,
              "character": 16
            }
          },
          "selectionRange": {
            "start": {
              "line": 31,
              "character": 9
            },
            "end": {
              "line": 31,
              "character": 16
            }
          },
          "children": []
        },
        {
          "name": "vroom2",
          "kind": 12,
          "range": {
            "start": {
              "line": 35,
              "character": 9
            },
            "end": {
              "line": 35,
              "character": 17
            }
          },
          "selectionRange": {
            "start": {
              "line": 35,
              "character": 9
            },
            "end": {
              "line": 35,
              "character": 17
            }
          },
          "children": []
        },
        {
          "name": "Car",
          "kind": 9,
          "range": {
            "start": {
              "line": 26,
              "character": 4
            },
            "end": {
              "line": 26,
              "character": 18
            }
          },
          "selectionRange": {
            "start": {
              "line": 26,
              "character": 4
            },
            "end": {
              "line": 26,
              "character": 18
            }
          },
          "children": []
        }
      ]
    },
    {
      "name": "IhaveAdefCtor",
      "kind": 5,
      "range": {
        "start": {
          "line": 40,
          "character": 0
        },
        "end": {
          "line": 41,
          "character": 1
        }
      },
      "selectionRange": {
        "start": {
          "line": 40,
          "character": 6
        },
        "end": {
          "line": 40,
          "character": 19
        }
      },
      "children": []
    }
  ],
  "symbols": {
    "34": {
      "loc": {
        "range": {
          "start": {
            "line": 2,
            "character": 4
          },
          "end": {
            "line": 2,
            "character": 15
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "THIS COMMENT USED #1"
    },
    "35": {
      "loc": {
        "range": {
          "start": {
            "line": 3,
            "character": 4
          },
          "end": {
            "line": 3,
            "character": 22
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "THIS COMMENT USED #2"
    },
    "36": {
      "loc": {
        "range": {
          "start": {
            "line": 16,
            "character": 4
          },
          "end": {
            "line": 16,
            "character": 9
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "x marks the spot\n  still line one still line one still line one still line one still line one\n  \n  line two\n  line two still\n \n  line three\n \n  - list item 1\n  - list item 2\n  - list item 3"
    },
    "3521": {
      "loc": {
        "range": {
          "start": {
            "line": 69,
            "character": 9
          },
          "end": {
            "line": 69,
            "character": 18
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "A lovely function."
    },
    "3525": {
      "loc": {
        "range": {
          "start": {
            "line": 26,
            "character": 4
          },
          "end": {
            "line": 26,
            "character": 18
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "a ctor with 1 param"
    },
    "3,0": {
      "loc": {
        "range": {
          "start": {
            "line": 23,
            "character": 8
          },
          "end": {
            "line": 23,
            "character": 13
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "How fast it goes"
    },
    "48": {
      "loc": {
        "range": {
          "start": {
            "line": 59,
            "character": 13
          },
          "end": {
            "line": 59,
            "character": 27
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "i am a car"
    },
    "41": {
      "loc": {
        "range": {
          "start": {
            "line": 55,
            "character": 13
          },
          "end": {
            "line": 55,
            "character": 23
          }
        },
        "uri": "ZQ_BUFFER"
      }
    },
    "39": {
      "loc": {
        "range": {
          "start": {
            "line": 73,
            "character": 8
          },
          "end": {
            "line": 73,
            "character": 11
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "hmmmmmm"
    },
    "37": {
      "loc": {
        "range": {
          "start": {
            "line": 18,
            "character": 4
          },
          "end": {
            "line": 18,
            "character": 5
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "hello world"
    },
    "38": {
      "loc": {
        "range": {
          "start": {
            "line": 18,
            "character": 7
          },
          "end": {
            "line": 18,
            "character": 12
          }
        },
        "uri": "ZQ_BUFFER"
      }
    },
    "32": {
      "loc": {
        "range": {
          "start": {
            "line": 48,
            "character": 4
          },
          "end": {
            "line": 48,
            "character": 8
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "absolute trash"
    },
    "31": {
      "loc": {
        "range": {
          "start": {
            "line": 46,
            "character": 4
          },
          "end": {
            "line": 46,
            "character": 9
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "top class"
    },
    "33": {
      "loc": {
        "range": {
          "start": {
            "line": 49,
            "character": 4
          },
          "end": {
            "line": 49,
            "character": 7
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "meh"
    },
    "3526": {
      "loc": {
        "range": {
          "start": {
            "line": 40,
            "character": 6
          },
          "end": {
            "line": 40,
            "character": 19
          }
        },
        "uri": "ZQ_BUFFER"
      }
    },
    "40": {
      "loc": {
        "range": {
          "start": {
            "line": 23,
            "character": 8
          },
          "end": {
            "line": 23,
            "character": 13
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "How fast it goes"
    },
    "3523": {
      "loc": {
        "range": {
          "start": {
            "line": 31,
            "character": 9
          },
          "end": {
            "line": 31,
            "character": 16
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "Go fast"
    },
    "4,0": {
      "loc": {
        "range": {
          "start": {
            "line": 23,
            "character": 8
          },
          "end": {
            "line": 23,
            "character": 13
          }
        },
        "uri": "ZQ_BUFFER"
      },
      "doc": "How fast it goes"
    },
    "54": {
      "loc": {
        "range": {
          "start": {
            "line": 26,
            "character": 8
          },
          "end": {
            "line": 26,
            "character": 17
          }
        },
        "uri": "ZQ_BUFFER"
      }
    }
  },
  "identifiers": [
    {
      "loc": {
        "line": 3,
        "character": 15,
        "length": 7
      },
      "symbol": "34"
    },
    {
      "loc": {
        "line": 56,
        "character": 26,
        "length": 7
      },
      "symbol": "34"
    },
    {
      "loc": {
        "line": 56,
        "character": 36,
        "length": 8
      },
      "symbol": "35"
    },
    {
      "loc": {
        "line": 56,
        "character": 50,
        "length": 1
      },
      "symbol": "36"
    },
    {
      "loc": {
        "line": 57,
        "character": 15,
        "length": 2
      },
      "symbol": "3521"
    },
    {
      "loc": {
        "line": 57,
        "character": 18,
        "length": 7
      },
      "symbol": "34"
    },
    {
      "loc": {
        "line": 59,
        "character": 21,
        "length": 3
      },
      "symbol": "3525"
    },
    {
      "loc": {
        "line": 60,
        "character": 32,
        "length": 5
      },
      "symbol": "3,0"
    },
    {
      "loc": {
        "line": 60,
        "character": 29,
        "length": 1
      },
      "symbol": "48"
    },
    {
      "loc": {
        "line": 60,
        "character": 39,
        "length": 6
      },
      "symbol": "41"
    },
    {
      "loc": {
        "line": 60,
        "character": 54,
        "length": 3
      },
      "symbol": "39"
    },
    {
      "loc": {
        "line": 60,
        "character": 60,
        "length": 1
      },
      "symbol": "37"
    },
    {
      "loc": {
        "line": 60,
        "character": 64,
        "length": 1
      },
      "symbol": "38"
    },
    {
      "loc": {
        "line": 60,
        "character": 68,
        "length": 4
      },
      "symbol": "32"
    },
    {
      "loc": {
        "line": 60,
        "character": 75,
        "length": 5
      },
      "symbol": "31"
    },
    {
      "loc": {
        "line": 60,
        "character": 83,
        "length": 3
      },
      "symbol": "33"
    },
    {
      "loc": {
        "line": 61,
        "character": 21,
        "length": 13
      },
      "symbol": "3526"
    },
    {
      "loc": {
        "line": 62,
        "character": 21,
        "length": 3
      },
      "symbol": "3525"
    },
    {
      "loc": {
        "line": 32,
        "character": 8,
        "length": 5
      },
      "symbol": "40"
    },
    {
      "loc": {
        "line": 36,
        "character": 8,
        "length": 5
      },
      "symbol": "3523"
    },
    {
      "loc": {
        "line": 27,
        "character": 14,
        "length": 5
      },
      "symbol": "4,0"
    },
    {
      "loc": {
        "line": 27,
        "character": 22,
        "length": 5
      },
      "symbol": "54"
    }
  ]
}

PUSHR d0; void run(int, int, int, int, int) Params
PUSHR d1
PUSHR d2
PUSHR d3
PUSHR d4
PUSHR REFFFC; void run(int, int, int, int, int) Body Start
PUSHARGSV 0,0.0005
SETR d4,SP2
ALLOCATEMEMV d2,7,0
STORE d2,0.0004
WRITEPODSTRING d2,"%d %d\n"
LOAD d2,0.0004; InlineFunc[void printf(char32)] Params Start
PUSHR d2
SETR d2,gd1
PUSHR d2
SETR d2,gd2
POP d3
ADDR d2,d3
PUSHVARGR d2
SETR d2,gd3
ADDV d2,1
ADDV d2,2
PUSHVARGR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0004
DEALLOCATEMEMR d3
PUSHR CLASS_THISKEY
PUSHR d4
PUSHV 2; ClassFunc[Car Car(int)] Params
CALLFUNC 0.0076; ClassFunc[Car Car(int)] Constructor Call
POP d4
POP CLASS_THISKEY
STORE_OBJECT d2,0.0003
ALLOCATEMEMV d2,10,0
STORE d2,0.0002
WRITEPODSTRING d2,"%d %d %d\n"
LOAD d2,0.0002; InlineFunc[void printf(char32)] Params Start
PUSHR d2
LOAD d2,0.0003
ZCLASS_READ d2,0
PUSHVARGR d2
LOAD d2,0.0010
PUSHVARGR d2
SETR d2,gd6
PUSHR d2
SETR d2,gd4
POP d3
ADDR d2,d3
PUSHR d2
SETR d2,gd5
POP d3
ADDR d2,d3
ADDV d2,1
ADDV d2,2
PUSHVARGR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0002
DEALLOCATEMEMR d3
SETV d2,0; ProtoFunc[IhaveAdefCtor IhaveAdefCtor()] Set Destructor
ZCLASS_CONSTRUCT d2,{0}; ProtoFunc[IhaveAdefCtor IhaveAdefCtor()] Default Construct
STORE_OBJECT d2,0.0001
PUSHR CLASS_THISKEY
PUSHR d4
PUSHV 2; ClassFunc[Car Car(int)] Params
CALLFUNC 0.0076; ClassFunc[Car Car(int)] Constructor Call
POP d4
POP CLASS_THISKEY
STORE_OBJECT d2,0
REF_REMOVE 0.0003
REF_REMOVE 0
REF_REMOVE 0.0001
POPARGS d5,0.0011
QUIT; void run(int, int, int, int, int) Body End
SETV d2,0; Func[Car Car(int)] Body Start
SETR CLASS_THISKEY2,CLASS_THISKEY
ZCLASS_CONSTRUCT CLASS_THISKEY,{0.0001}
SETR d4,SP2
PUSHR CLASS_THISKEY2
LOAD d2,0
PUSHR d2
SETR d2,CLASS_THISKEY
SETR d3,d2
POP d2
ZCLASS_WRITE d3,0
POP d5
SETR d2,CLASS_THISKEY
POP CLASS_THISKEY
RETURNFUNC; Func[Car Car(int)] Body End


SETV gd1,1; void run() Body Start
SETV gd2,1
SETV gd3,1
SETV gd4,0
SETV gd5,2
SETV gd6,0
QUIT; void run() Body End